sentry/patch.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2022 Akretion (https://www.akretion.com).
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+
+# Monkey patch odoo method in order to track all database
+
+
+import logging
+
+import psycopg2
+
+import odoo.http
+from odoo.sql_db import Cursor
+
+from odoo.addons.base.models.ir_cron import ir_cron
+
+_logger = logging.getLogger(__name__)
+HAS_SENTRY_SDK = True
+
+try:
+    from sentry_sdk import start_transaction
+    from sentry_sdk.hub import Hub
+    from sentry_sdk.tracing_utils import record_sql_queries
+except ImportError:  # pragma: no cover
+    HAS_SENTRY_SDK = False  # pragma: no cover
+    _logger.debug(
+        "Cannot import 'sentry-sdk'.\
+                        Please make sure it is installed."
+    )  # pragma: no cover
+
+
+if HAS_SENTRY_SDK:
+
+    _ori_get_request = odoo.http.Root.get_request
+
+    def get_request(self, httprequest):
+        request = _ori_get_request(self, httprequest)
+        hub = Hub.current
+        with hub.configure_scope() as scope:
+            # Extract some params of the request to give a better name
+            # to the transaction and also add tag to improve filtering
+            # experience on sentry
+            scope.set_transaction_name(httprequest.environ.get("PATH_INFO"))
+            scope.set_user({"id": httprequest.session.get("uid")})
+            for key in ["model", "method"]:
+                if key in request.params:
+                    scope.set_tag(f"odoo.{key}", request.params[key])
+            scope.set_tag("odoo.db", request.db)
+        return request
```

Comment:
I'm wondering why we can't reuse the [before_send](https://github.com/OCA/server-tools/blob/14.0/sentry/hooks.py#L36) (which is calling [get_extra_context](https://github.com/OCA/server-tools/blob/14.0/sentry/logutils.py#L30))  mechanism to retrieves such info instead mockey patching get_request ? probably some variable or not set there or you missed this behavior ?

Reply:
I running to this error when using with OCA/rest-framework

```python
            [UTC_TIME] ERROR [DB_NAME] werkzeug: Error on request:
            Traceback (most recent call last):
            File ".../.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 306, in run_wsgi
                execute(self.server.app)
            File ".../.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 294, in execute
                application_iter = app(environ, start_response)
            File ".../odoo/custom/src/odoo/odoo/service/server.py", line 482, in app
                return self.app(e, s)
            File ".../odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 112, in application
                return application_unproxied(environ, start_response)
            File ".../odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 87, in application_unproxied
                result = odoo.http.root(environ, start_response)
            File ".../odoo/custom/src/odoo/odoo/http.py", line 1341, in __call__
                return self.dispatch(environ, start_response)
            File ".../odoo/custom/src/odoo/odoo/http.py", line 1307, in __call__
                return self.app(environ, start_wrapped)
            File ".../.venv/lib/python3.9/site-packages/werkzeug/middleware/shared_data.py", line 220, in __call__
                return self.app(environ, start_response)
            File ".../odoo/custom/src/odoo/odoo/http.py", line 1496, in dispatch
                request = self.get_request(httprequest)
            File ".../odoo/custom/src/server-tools/sentry/patch.py", line 39, in get_request
                request = _ori_get_request(self, httprequest)
            File ".../odoo/custom/src/rest-framework/base_rest/http.py", line 234, in get_request
                return ori_get_request(self, httprequest)
            File ".../odoo/custom/src/server-tools/sentry/patch.py", line 39, in get_request
                request = _ori_get_request(self, httprequest)
            File ".../odoo/custom/src/rest-framework/base_rest/http.py", line 230, in get_request
                """ :class:`~collections.abc.Mapping` of context values for the current request """
            NameError: name '_rest_services_routes' is not defined - - -
```
basically, I think that 2 patching using pointer reset like this running into 1 issue, some variables defined in scope 1 (`base_rest/http.py`) is not yet define in scope 2 (`sentry/patch.py`)